os/drivers/usbhost/uvc/usbhost_uvc_ctrl.c

```diff
@@ -0,0 +1,1360 @@
+		} else {
+			ret = uvc_query_ctrl(chain->dev, UVC_GET_CUR, ctrl->entity->id, chain->dev->intfnum, ctrl->info.selector, ctrl->cur, ctrl->info.size);
+			if (ret < 0) {
+				udbg("Failed to query the control(%d) to device!!\n", ctrl->info.selector);
+				return ret;
+			}
+			ret = OK;
+		}
+
+		CTRL_SET_LOADED(ctrl->status);
+	}
+
+	/* Backup the current value in case we need to rollback later. */
+	if (!CTRL_IS_DIRTY(ctrl->status)) {
+		memcpy(ctrl->bak, ctrl->cur, ctrl->info.size);
+	}
+
+	uvc_set_value(mapping, value, ctrl->cur);
+
+	CTRL_SET_DIRTY(ctrl->status);
+	CTRL_SET_MODIFIED(ctrl->status);
+	return 0;
+}
+
+/****************************************************************************
+ * Name: uvc_ctrl_restore_values
+ *
+ * Description:
+ *   This function restore the value on resume from a low power cycle.
+ *
+ *   This function is called from the resume method of driver.
+ *
+ * Input Parameters:
+ *   dev - reference to the driver object.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+int uvc_ctrl_restore_values(uvc_state_t *dev)
+{
+	struct uvc_control *ctrl;
+	struct uvc_entity *entity = queue_entry(dev->entities.head, struct uvc_entity, list);
+	unsigned int i;
+	int ret = OK;
+
+	/* Walk the entities list and restore controls when possible. */
+	for (; &entity->list != (dq_entry_t *)&dev->entities; entity = queue_entry(entity->list.flink, struct uvc_entity, list)) {
+
+		for (i = 0; i < entity->ncontrols; ++i) {
+			ctrl = &entity->controls[i];
+
+			if (!CTRL_IS_SUPPORTED(ctrl->status) || !CTRL_IS_MODIFIED(ctrl->status) || (ctrl->info.flags & F_RESTORE) == 0) {
+				continue;
+			}
+
+			uvdbg("restoring control %pUl/%u/%u\n", ctrl->info.entity, ctrl->info.index, ctrl->info.selector);
+			CTRL_SET_DIRTY(ctrl->status);
+
+			if (ctrl->info.flags & F_AUTO || !(ctrl->info.flags & F_GET_CUR)) {
+				CTRL_CLEAR_LOADED(ctrl->status);
+			}
+
+			ret = uvc_query_ctrl(dev, UVC_SET_CUR, ctrl->entity->id, dev->intfnum, ctrl->info.selector, ctrl->cur, ctrl->info.size);
+
+			if (ret < 0) {
+				memcpy(ctrl->cur, ctrl->bak, ctrl->info.size);
+			} else {
+				ret = OK;
+			}
+
+			CTRL_CLEAR_DIRTY(ctrl->status);
+		}
+	}
+
+	return ret;
+}
+
+/****************************************************************************
+ * Name: allocate_ctrl_vals
+ *
+ * Description:
+ *   This function allocate memory for min/max/default/current/backup.
+ *
+ *   This function is called from the uvc_ctrl_add_info method.
+ *
+ * Input Parameters:
+ *   ctrl - reference to the control object.
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failures
+ *
+ ****************************************************************************/
+static int allocate_ctrl_vals(struct uvc_control *ctrl)
+{
+	int ret = OK;
+	/* Allocate memory for each of the controls */
+	ctrl->min = kmm_zalloc(ctrl->info.size);
```

Comment:
we don't have to null check for ctrl->min, max, cur, step  here? if it initiated as a null when entity initialized, please ignore this comment.

Reply:
ctrl is zero allocated so min, max, cur step will be zero initialized.

Reply:
ah. ok got it.